### PR TITLE
Output hidden states

### PR DESCRIPTION
**Pull request by drjosephliu**
Attempts to close issue #3879 by refactoring all models to take in an extra argument `output_hidden_states` in the `forward()` method. 

@patrickvonplaten

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[[source]]
```
Comment (patrickvonplaten) [inline Pipfile:1]: Can we delete this file? :-)

**Comment by patrickvonplaten** [comment]
@drjosephliu - this is really great work! I added two additional tests for output hidden states that have to pass for all models. Can you fix the remaining models? :-) I think after that we are good to merge!

**Comment by julien-c** [comment]
This is great, looking forward to having this in master:)

**Comment by drjosephliu** [comment]
Just pushed the fixes. Fingers crossed for a hole in one here !

**Comment by sshleifer** [comment]
Looks like one more merge conflict :(. Otherwise LGTM, epic contribution!

**Comment by drjosephliu** [comment]
I added the refactorisation for mobilebert, so the relevant tests should be passing now. However, there are a bunch of tokenisation tests which are now failing instead and I can't figure out why as I'm pretty certain the changes I made didn't affect the tokenisers whatsoever. FWIW, those tests were already failing on master, so I hope they're unrelated to the changes I've made. Anyways, let me know if there are any additional edits that need to be made.

P.S. what a coincidence to bump into you here @sshleifer ! Glad to see you settling in here.

**Comment by sshleifer** [comment]
Haha small world!

I think there is a git issue causing it to appear that 188 files are changed in the PR.

I don't think it's horribly damaging, (the RHS looks wrong, the LHS looks correct  in the diff viewer), but if you have an easy way to resolve it that would be nice.

File: Pipfile (hunk 1 above)
Comment (sshleifer) [reply Pipfile:1]: doesnt seem resolved

**Comment by drjosephliu** [comment]
Well crap, it seems like it's gotten worse. I'm out of ideas here, because I've `git fetch upstream` and `git merge upstream/master`, so it's telling me everything's up to date. It seems like it's comparing against an old version of master that's a couple commits ago. Perhaps I can try deleting my local master, creating and pulling a new master, merging output_hidden_states into master and then push and make a new PR. I'm wondering if that will work? Unless you have any other ideas.

**Comment by drjosephliu** [comment]
Ok, so with a bit of git ninja'ing, it's no longer showing ~200 files changed. Hoping this works now.

File: Pipfile (hunk 1 above)
Comment (drjosephliu) [reply Pipfile]: Should be gone now.

**Review by sshleifer — APPROVED** [review]
LGTM from my end @patrickvonplaten !

**Comment by patrickvonplaten** [comment]
@drjosephliu - Amazing work! Thank's a lot for this. Really helps the library to become more flexible :-) 

The PR is good to merge for me.

Pinging @LysandreJik to verify since it's a big one. Should be merged today though I hope :-)

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -517,6 +511,18 @@ def test_hidden_states_output(self):
...
+
+        for model_class in self.all_model_classes:
+            inputs_dict["output_hidden_states"] = True
+            check_hidden_states_output(inputs_dict, config, model_class)
+
+            # check that output_hidden_states also work using config
+            del inputs_dict["output_hidden_states"]
+            config.output_hidden_states = True
+
+            check_hidden_states_output(inputs_dict, config, model_class)
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:524]: Nice tests!

**Review by LysandreJik — APPROVED** [review]
This looks great, thanks a lot!

**Comment by drjosephliu** [comment]
Awesome, no problem and great working with y'all !

**Pull request merged**